### DISCUSSION TITLE
Why preact/compat ignoring the selected attribute of a option?

### DISCUSSION THREAD

**Discussion by nihil-pro** [preactjs/preact#4467] [Q&A]
I have an custom element, which uses options as children. But, when i pass the selected attribute, result of option.getAttribute('selected') is null, and also, the MutationObserver didn't react on this attribute change

**Comment by rschristian** [comment]
Can you please provide a reproduction? There's not much we can say on the matter from your description alone.

**Reply by nihil-pro** [reply]
Sure. [Sandbox](https://codesandbox.io/p/sandbox/option-selected-jv6xqn)

```
function App() {
  return (
    <select>
      <option value={123} selected> // under the hood of preact, 'selected' will be set as domElement.selected = true, instead of domElement.setAttribute('selected', '')
        Text
      </option>
    </select>
  );
}
```

Because of this, we can't observe select.children.attributes via [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver/observe), i.e. observer.observe(selectRef, { filterAttributes: ['selected'], subTree: true })

**Reply by rschristian** [reply]
I think you've misunderstood what the `selected` attribute is for -- it's for specifying a _default selected_ option, it will not reflect the currently selected option nor is it meant to.

> [selected](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option#selected)
>
>    If present, this Boolean attribute indicates that the option is initially selected. If the <option> element is the descendant of a [<select>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select) element whose [multiple](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#multiple) attribute is not set, only one single <option> of this [<select>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select) element may have the selected attribute.

[MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/option#selected)

Here's an [REPL example](https://preactjs.com/repl?code=ZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQoJ2FwcCcpLmlubmVySFRNTCA9IGAKCTxzZWxlY3Q%2BCgkJPG9wdGlvbiB2YWx1ZT0iMTIzIiBzZWxlY3RlZD5UZXh0PC9vcHRpb24%2BCgkJPG9wdGlvbj5UZXh0Mjwvb3B0aW9uPgoJPC9zZWxlY3Q%2BCmA7Cgpkb2N1bWVudC5xdWVyeVNlbGVjdG9yKCdzZWxlY3QnKS5hZGRFdmVudExpc3RlbmVyKCdjaGFuZ2UnLCAoKSA9PiB7Cgljb25zdCBvcHRpb25zID0gZG9jdW1lbnQucXVlcnlTZWxlY3RvckFsbCgnb3B0aW9uJyk7Cgljb25zb2xlLmxvZyhvcHRpb25zWzBdLmdldEF0dHJpYnV0ZSgnc2VsZWN0ZWQnKSk7Cgljb25zb2xlLmxvZyhvcHRpb25zWzFdLmdldEF0dHJpYnV0ZSgnc2VsZWN0ZWQnKSk7Cn0pOw%3D%3D), using plain HTML.

Yes, Preact does set the property rather than the attribute, but watching for attribute changes isn't the correct way to go about this as it's not how `<select>`/`<option>` works. If #4419 lands, you'll have a syntax for specifying props vs attributes, but you should find another way to do what it is you're after.

Are you just trying to watch for changes? If so, putting a `onchange` handler onto the `<select>` would be a good option, else, you can always use data attributes (`data-my-selected`).

**Reply by nihil-pro** [reply]
I don't agree. There is a difference between HTMLElement properties and attributes, and selected is clearly attribute [Impl](https://github.com/chromium/chromium/blob/main/third_party/blink/renderer/core/html/forms/html_option_element.cc).
Also, the [Select](https://github.com/chromium/chromium/blob/main/third_party/blink/renderer/core/html/forms/html_select_element.cc) checking the selected attribute, not property.

**Reply by rschristian** [reply]
There is a difference, sure, and you're mixing up the use for the attribute and the use for the property.

The property is live, a reflection of the currently selected option. It changes when you select a different option.
The attribute is static, only meant to specify the initial value. It does not change without JS DOM manipulation.

I can't speak to Chrome code, I'm not at all familiar with it (nor did you link to any specific lines so I'm not sure what in the world I'm supposed to spot in those 2400 lines of C++), but this is very easy to verify by just looking at the DOM. [Updated REPL example](https://preactjs.com/repl?code=ZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQoJ2FwcCcpLmlubmVySFRNTCA9IGAKCTxzZWxlY3Q%2BCgkJPG9wdGlvbiB2YWx1ZT0iMTIzIiBzZWxlY3RlZD5UZXh0PC9vcHRpb24%2BCgkJPG9wdGlvbj5UZXh0Mjwvb3B0aW9uPgoJPC9zZWxlY3Q%2BCmA7Cgpkb2N1bWVudC5xdWVyeVNlbGVjdG9yKCdzZWxlY3QnKS5hZGRFdmVudExpc3RlbmVyKCdjaGFuZ2UnLCAoKSA9PiB7Cgljb25zdCBvcHRpb25zID0gZG9jdW1lbnQucXVlcnlTZWxlY3RvckFsbCgnb3B0aW9uJyk7Cgljb25zb2xlLmxvZyhvcHRpb25zWzBdLmdldEF0dHJpYnV0ZSgnc2VsZWN0ZWQnKSwgb3B0aW9uc1swXS5zZWxlY3RlZCk7Cgljb25zb2xlLmxvZyhvcHRpb25zWzFdLmdldEF0dHJpYnV0ZSgnc2VsZWN0ZWQnKSwgb3B0aW9uc1sxXS5zZWxlY3RlZCk7Cn0pOw%3D%3D).

**Reply by nihil-pro** [reply]
This is the browser default behavior. We see in dom attribute, not property. Yes, if we select option 2, the selected attribute of option 1 remains in the dom. 
<img width="880" alt="Screenshot 2024-08-03 at 10 45 45" src="https://github.com/user-attachments/assets/59ba6722-da38-4895-9785-87cef5c734a0">

If we use preact, selected will be a property. 
<img width="1094" alt="Screenshot 2024-08-03 at 10 48 51" src="https://github.com/user-attachments/assets/67ac3fe9-ab96-4ba4-b9ba-7281767639e8">

**Reply by nihil-pro** [reply]
I just want the same behavior. 
Oh, and it's funny, that if i pass 'Selected', it works as expected. Probably it's treated as custom attribute

**Reply by rschristian** [reply]
In one you're writing HTML, in the other JS. It doesn't make much sense for a JS framework to set the `selected` attribute (outside of SSR, which we do support & do).

As mentioned, if #4419 lands it will allow this, else, you'll need to use refs.

Edit:
>  Oh, and it's funny, that if i pass 'Selected', it works as expected. Probably it's treated as custom attribute

In Preact, outside of a few excluded names, anything found on the DOM element (`'Selected' in document.createElement('div')`, essentially) is set as a prop, anything else is an attribute. As the prop is `selected`, `Selected` fails this check and becomes an attribute instead.

**Reply by JoviDeCroock** [reply]
You can also learn more about this distinction in here https://jakearchibald.com/2024/attributes-vs-properties/ but it's one of the hard choices we as library authors have to make. When we send html over the wire with i.e. renderToString it will all be attributes and then be rehydrated and highly interactive things will be properties.

> In that sense, I think `<input value>` gets it right (aside from the naming). The value attribute configures the default value, whereas the value property gives you the current state.

One larger downside of attributes is that even when you change them to the same value, it will be counted as a mutation 😅

**Reply by nihil-pro** [reply]
Ok. I'm using the data attribute now, but I think you're wrong. That's what I mean. The 'selected' should be set as attribute, just like in the browser. The fact is that when selecting an option in a select, regardless of the multiple attribute, and regardless of the presence of the selected attribute, the selected property will be set to true via browser, and vice versa. 

Just try this
```
<select multiple>
  <option selected>1</option> // option.selected = true, option.getAttribute('selected') = empty string
  <option >2</option> // option.selected = false, option.getAttribute('selected') = null
  <option>3</option>
</select>
```

```
document.querySelector('select').addEventListener('change', e => {
  event.target.querySelectorAll('option').forEach(o => {
    console.log('attr', o.textContent, o.getAttribute('selected'))
    console.log('prop', o.textContent, o.selected)
  })
})
document.querySelector('select').querySelectorAll('option').forEach(o => {
    console.log('attr', o.textContent, o.getAttribute('selected'))
    console.log('prop', o.textContent, o.selected)
  })
```

**Reply by nihil-pro** [reply]
> One larger downside of attributes is that even when you change them to the same value, it will be counted as a mutation 😅

I know. And I expect this behavior from the attribute selected for the option both inside the select and inside the datalist

**Reply by JoviDeCroock** [reply]
> Ok. I'm using the data attribute now, but I think you're wrong. That's what I mean. The 'selected' should be set as attribute, just like in the browser.

Before you say this I really think you should read yourself into this knowledge 😅 these blanket statements of you're wrong are really offensive and talk down on the wealth of knowledge we as maintainers have to gather to even engage with these subjects let alone create a library like this.

I am stepping away from this conversation as you don't really want to discuss the differences between JS/HTML either way nor do you want to find the middle ground. I am happy to improve this library when it's appropriate but comparing everything to the HTML equivalent feels a bit like you're trying to dodge the conversation and pushing your opinion.

**Reply by nihil-pro** [reply]
> > Ok. I'm using the data attribute now, but I think you're wrong. That's what I mean. The 'selected' should be set as attribute, just like in the browser.
> 
> Before you say this I really think you should read yourself into this knowledge 😅 these blanket statements of you're wrong are really offensive and talk down on the wealth of knowledge we as maintainers have to gather to even engage with these subjects let alone create a library like this.
> 
> I am stepping away from this conversation as you don't really want to discuss the differences between JS/HTML either way nor do you want to find the middle ground. I am happy to improve this library when it's appropriate but comparing everything to the HTML equivalent feels a bit like you're trying to dodge the conversation and pushing your opinion.

I use a translator because I don't speak English. There was no attempt to offend in my message, I apologize if this was understood that way.

**Reply by rschristian** [reply]
> The 'selected' should be set as attribute, just like in the browser.

Preact runs in the browser, so I'm not quite sure what you're referring to.

HTML and the DOM are two different interfaces that have some overlap; they work differently. Preact, as it's JS, works with DOM, and therefore properties, primarily. We work in the same way as vanilla JS here. The `selected` attribute is not reflected as it's not meant to show the currently selection option, but an initial value.

If you want to set an initial value only (as you still haven't clarified use), there is the `defaultSelected` property which reflects to the `selected` attribute?

**Reply by nihil-pro** [reply]
I understand difference between html and js. I want to say, that if preact will set selected as an attribute, nothing brokes, because:
```
const select = document.createElement('select')
const option = document.createElement('option')
option.setAttribute('selected', '')
select.appendChild(option)
document.documentElement.appendChild(select) 
// after append, option.selected will be true according to the presence of the selected attribute
```

**Reply by rschristian** [reply]
I don't think this conversation is productive.

We work primarily with DOM properties, as do nearly all JS frameworks. This behavior is entirely intended.

If you need the `selected` attribute, there's a number of workarounds that have been suggested from which you'll need to pick one. Good luck.

**Comment by nihil-pro** [comment]
One more question please. 
A little problem with dreggable attribute. 
`Warning: This attribute is enumerated  and not Boolean. A value of true or false is mandatory, and shorthand like <img draggable> is forbidden. The correct usage is <img draggable="false">. `[MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/draggable) 
But, if I set it like draggable="false", preact sets it like draggable="true", if I set it like draggable={false} preact sets it like draggable="false". Is this done intentionally?

**Reply by JoviDeCroock** [reply]
Same answer as https://github.com/preactjs/preact/issues/4464#issuecomment-2261680451 - you are alluding to the draggable _attribute_ in that quote not to the draggable _property_.

https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/draggable vs https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/draggable

**Reply by rschristian** [reply]
And unfortunately, attribute reflection can be a bit questionable at times. Preact sets _the property_ to `"false"`, and because it's meant to be a boolean, this is truthy and therefore reflected to the attribute as `"true"`.

It's how the DOM works, isn't related to Preact really:

```js
const div = document.createElement('div');
div.draggable = "false";
console.log(div.getAttribute('draggable')); // true
```

**Reply by nihil-pro** [reply]
With this and [this](https://github.com/preactjs/preact/blob/main/src/diff/props.js#L101) in mind, it might be helpful to include  in the [documentation](https://preactjs.com/guide/v10/differences-to-react#raw-html-attributeproperty-names) something like this: "Note that only the aria and data attributes are set as attributes. Everything else is set as properties, depending on the type of HtmlElement instance property." What do you think about it?

**Reply by rschristian** [reply]
I would say that's not very accurate, see:

https://github.com/preactjs/preact/blob/91679e5f4d06087cafe943b779c53cbdc8633b30/src/diff/props.js#L107-L120
